cli/utils.go

```diff
@@ -0,0 +1,151 @@
+package cli
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"sort"
+	"strings"
+
+	"github.com/spf13/viper"
+
+	"github.com/prometheus/alertmanager/pkg/parse"
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/pkg/labels"
+)
+
+type ByAlphabetical []labels.Matcher
+
+func (s ByAlphabetical) Len() int      { return len(s) }
+func (s ByAlphabetical) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
+func (s ByAlphabetical) Less(i, j int) bool {
+	if s[i].Name != s[j].Name {
+		return s[i].Name < s[j].Name
+	} else if s[i].Type != s[j].Type {
+		return s[i].Type < s[j].Type
+	} else if s[i].Value != s[j].Value {
+		return s[i].Value < s[j].Value
+	} else {
+		return false
+	}
+}
+
+func GetAlertmanagerURL() (*url.URL, error) {
+	u, err := url.ParseRequestURI(viper.GetString("alertmanager.url"))
+	if err != nil {
+		return nil, errors.New("Invalid alertmanager url")
+	}
+	return u, nil
+}
+
+// Parse a list of labels (cli arguments)
+func parseMatchers(inputLabels []string) ([]labels.Matcher, error) {
+	matchers := make([]labels.Matcher, 0)
+
+	for _, v := range inputLabels {
+		fmt.Println(v)
+		matcher, err := parse.Matchers(v)
+		if err != nil {
+			return []labels.Matcher{}, err
+		}
+		for _, item := range matcher {
+			matchers = append(matchers, *item)
+		}
+	}
+
+	return matchers, nil
+}
+
+// Expand a list of matchers into a list of all combinations of matchers for each matcher that has the same Name
+// ```
+// alertname=foo instance=bar instance=baz
+// ```
+// Goes t
+// ```
+// alertname=foo instance=bar
+// alertname=foo instance=baz
```

Comment:
In this example, what's the reason for splitting this into multiple silences with a single `instance` label instead of a single silence with multiple instance labels? `alertname=foo instance=bar instance=baz` affects the same alerts, and produces (to me) the expected result from using the command line tool. I would be surprised if, for instance, this label set:

```
alertname=fooService alertname=bazService statusCode=400 statusCode=500
```

produced four silences:

```
alertname=fooService  statusCode=500
alertname=fooService statusCode=400
alertname=bazService statusCode=500
alertname=bazService statusCode=400
```

So I'm just curious what the use-case is that this is solving?